aten/src/ATen/native/Sorting.cpp

```diff
@@ -94,6 +91,169 @@ void quick_select_template(
   } while (true);
 }
 
+// Returns pointer to pivot used to partition range [first, last]
+// There should be at least 3 elements in [first, last]
+template <typename random_it, typename comp>
+random_it partition(random_it first, random_it last, comp lt) {
+  // Use median of three for pivot choice
+  random_it pivot = first + (std::distance(first, last) >> 1);
+  std::swap(first[1], *pivot);
+  if (lt(*last, first[1])) {
+    std::swap(first[1], *last);
+  }
+  if (lt(*last, *first)) {
+    std::swap(*first, *last);
+  }
+  if (lt(*first, first[1])) {
+    std::swap(first[1], *first);
+  }
+  pivot = first++;
+
+  do {
+    do {
+      ++first;
+    } while (lt(*first, *pivot));
+    do {
+      --last;
+    } while (lt(*pivot, *last));
+    if (last < first) {
+      break;
+    }
+    std::swap(*first, *last);
+  } while (true);
+  std::swap(*pivot, *last);
+
+  return last;
+}
+
+template <typename random_it, typename kth_iterator, typename comp>
+void quickselect_helper(
+    random_it start,
+    random_it first,
+    random_it last,
+    kth_iterator kbegin,
+    kth_iterator kend,
+    comp lt) {
+  // Base case: one element or no kth values
+  if (last <= first || kbegin == kend) {
+    return;
+  }
+
+  // Base case: two elements
+  if (last == first + 1) {
+    if (lt(*last, *first)) {
+      std::swap(*first, *last);
+    }
+    return;
+  }
+
```

Comment:
To be parallel with previous comments, this would have a comment like "General case: 3 or more elements"